### PR TITLE
[bug-fix] hparams not plotting metrics correctly

### LINKED ISSUE #37738: Allow user to update metrics in Tensorboard SummaryWriter.add_hparam
## Motivation

Currently, every call to `SummaryWriter.add_hparam` creates a new `SummaryWriter` which retains a separate record of the metrics associated with the given set of parameters. This makes it impossible to update the metric associated with a given set of hyperparameters. For example, I would like to do

```python
writer = SummaryWriter()
writer.add_hparams({"lr" : 0.001}, {"accuracy" : 0})
for epoch in range(epochs):
    ...
    writer.add_scalar("accuracy", get_accuracy(epoch), epoch)
```

and have the accuracy in the hparam console track the current "accuracy" scalar. Right now this fails because `add_hparams` creates a new directory with its own accuracy scalar that is always 0. This makes the "Show Metrics" button in Tensorboard useless, and it's also unexpected behavior because you would not expect a `SummaryWriter` to create a new `SummaryWriter` internally with a new directory for every parameter. If this behavior were changed, you could have results like [this](https://imgur.com/a/Z5u2Fer).

## Proposal

This is not trivial behavior to change. Tensorflow handles this by supporting

```python
  with tf.summary.create_file_writer(run_dir).as_default():
    hp.hparams(hparams)  # record the values used in this trial
    accuracy = train_test_model(hparams)
    tf.summary.scalar(METRIC_ACCURACY, accuracy, step=1)
```

but PyTorch's summary writer does not have a notion of `as_default()`, and it's not clear that `add_hparams` rightly belongs as a member function in `SummaryWriter`. I see three ways to support this.

1. Have `add_hparams` return the new `SummaryWriter` that it used to write the summary. This could be ignored if not needed but otherwise it could be used to update metrics during a training run. This has the downside of returning writers that need to be closed. A `return_writer=False` flag could be added to mitigate this.
2. Add a `use_current_writer=False` flag to `add_hparams` that would use the default file writer if `True` and track global metrics instead of creating a new `SummaryWriter`. This would be disabled by default, but would allow users to track global metrics if desired. 
3. Add a `replace_writer` flag to `add_hparams` that replaces the `file_writer` in the `SummaryWriter` with the new writer created by `add_hparams`. This would be False by default, but could be set to True if you want the metrics to track future calls to `add_scalar`.

I think the first is preferable, although it still pollutes the log directory with new files for every call to `add_hparams`. The third is a bit cleaner with respect to cleaning up file pointers, but it's more confusing from a user perspective.

## Code changes

First approach:

`torch/utils/tensorboard/writer.py:305`
```python
w_hp = SummaryWriter(log_dir=logdir)
w_hp.file_writer.add_summary(exp)
w_hp.file_writer.add_summary(ssi)
w_hp.file_writer.add_summary(sei)
for k, v in metric_dict.items():
    w_hp.add_scalar(k, v)

if return_writer:
    return w_hp
else:
    w_hp.close()
```

## Additional context

This is the kind of metric view you can access if this is supported.

Screen Shot 2020-05-03 at 2 24 16 PM

Currently, this is always a single value, i.e.

Screen Shot 2020-05-03 at 2 38 04 PM

### PR DESCRIPTION
**Pull request by dreizehnutters**
This is similar to tensorboardX implementation.

Fixes #37738

### CODE REVIEW

**Comment by nlgranger** [comment]
The stale label is wrong, the bug this PR fixes still exists.

**Comment by dreizehnutters** [comment]
Thanks for noticing... I don't know how to get a reviewer for the PR.

**Comment by vwxyzjn** [comment]
Hey just wondering if there is any support in this. This feature and PR feel like a quick fix and should be supported.

**Comment by Ryul0rd** [comment]
This shouldn't be closed as the issue still exists. @ngimel it seems like J0Nreynolds isn't going to review this so could you assign someone else?

**Comment by ngimel** [comment]
Please rebase the PR and I'll land it

**Comment by dreizehnutters** [comment]
i am no git pro. id did a `git rebase -i` in the pr branch and pushed it again. hope it helps

**Comment by ringohoffman** [comment]
Are there any plans to release this fix? It's still not in [v2.1.0-rc5](https://github.com/pytorch/pytorch/blob/ced78cc2a797229c3ebe2096e7e6d52ef3911221/torch/utils/tensorboard/writer.py#L295).
